### ISSUE TITLE
Use of "inbox"

### ISSUE THREAD

**Issue by klensin** [csarven/ldn#52]
The term "inbox" has a very long history (long before there was a web) in the context of email.  It carries very strong semantics, especially with regard to what mail delivery servers (MDAs in popular email terminology) and Mail User Agents (MUAs), including so-called split ones (e.g., POP3 and IMAP4 servers and clients) are allowed tor expected to do as well as a sometimes-precise relationship to a "mail store".  This is especially important when internationalization is involved because those possible actions are expected to include language-selection or even automatic translation functions.  While its primary use in recent years has been to allow plain-text and HTML-formatted versions of a message to be sent together, the multipart/alternative content type was introduced precisely to allow a mail message to be sent with versions in several different languages, with the preferred one selected by the user or her agent.

The relationship between an inbox and a mail store further complicates this LDN use because it is plausible to expect that, in some implementations, the same object storage mechanism might be used for email inboxes (or email message collections from while virtual inboxes can be constructed or views or on demand) and for LDN messages.    One can also imagine the use of LDN to get "you have mail" or equivalent  notifications to a user, further increasing to confusion between an LDN inbox and an email inbox.

The use of that term with rather different semantics and operations in the LDN context and the email context  is an invitation for considerable implementer confusion and far greater user confusion when it leaks into user-level terminology and documentation.   It, and this type of reuse of long-established and made-up, and therefore application-specific, terms more generally, should be avoided.

"depository", "container", "store" (perhaps with some prefix term) or some other term that does not have a specific and well-established meaning on the Internet would be more appropriate.

**Comment by cwebber** [comment]
Inbox's use here is pretty clearly defined though.  It's also used in ActivityPub (and are sharing the same definition of inbox), where there is both inbox and outbox, and that is pretty deeply baked and isn't going away.

So while "inbox" may have a lot of different meanings elsewhere, it has a fairly particular meaning here.  You can also say the same thing about "actor" in ActivityPub... the field of the [actor model](https://en.wikipedia.org/wiki/Actor_model) is a well-researched area, and while there's strong overlap, it's not be used exactly in that way.  What "Actor" means for ActivityPub is clearly defined for its scope.  Like "inbox", even though there are other meanings that can be viewed out there, including in CS, as in terms of our nomenclature, we've carefully defined it.  (Isn't that the goal of namespaces?)

So I don't see what the problem is?

**Comment by aphillips** [comment]
I don't agree with this issue. While the name "Inbox" would not have been my first choice (for reasons similar to @klensin), we are talking about a spec at CR and with implementations. Changing the name to 'dumpster' or 'basket' or 'social-tarball' or whatnot seems arbitrary and capricious to those who have been working on it.

**Comment by klensin** [comment]
Addison and Christopher,
Ok.  Suppose we assume, for purposes of discussion, that every reader of this document will notice the local definition or "inbox", apply it, and have no difficulties keeping track of that definition (after a few years of experience with specifications and implementers, I don't believe that, but let's assume it and move on).  I encourage you to think about one example/issue that your notes do not address and wonder what implementations of either LDR or, if applicable, ActivityPub, are doing about it   Suppose an email message is received for a given user, placed in the relevant mail store, a LDR used to generate a notification of its arrival.  Then something goes wrong.   What does the error message say, given that "a message in your inbox was frobbed" is hopelessly ambiguous as to whether it is about the mail message or the notification, independent of whatever technical jargon is substituted for "frobbed".

If you are confident that there are good answers for that sort of situation, I drop my objection.   If not, I suggest that the purpose of these reviews is to uncover glitches before there is a final document and that this is a glitch that needs to be addressed.  I don't have a strong opinion whether, given the circumstances, it is best addressed by changing the terminology., by making some specific suggestions about error identification terminology, or something else.

**Comment by cwebber** [comment]
@klensin I think you're anticipating that "inbox" is a much more user-facing detail than any of the rest of us are?  In ActivityPub, most users won't even know they have an "inbox" endpoint... it's not relevant to them, it's relevant to people implementing the protocol.

As a pump.io user which uses "inbox" and "outbox", I spent the first year of using it not even aware that it was using endpoints with those names under the hood!

**Comment by klensin** [comment]
@Christopher:   I don't understand ActivityPub nearly well enough to understand the relationship between terminology in the spec and what the user sees there but, if that answer is close to "none", as you comment above implies, that would not surprise me at all.  However, with LDN. , it seems to me that there is an obvious error state (see prior message).  There is a very long history -- with the Internet going back to the beginning and with software in general -- of problems at or near the application layer getting into error states and, because those states and how to handle them are explicit parts of the specification,, the resulting messages appearing with no usable information at all, information that leaks terminology that was expected to appear only in the specification, that was expected to be buried in the implementation and never be seen by users, or both.    Of course, if errors never occur, that issue is a non-problem.

**Comment by csarven** [comment]
@klensin Thanks for raising this. As mention in the call, I am in agreement with your general concern.

I think the notification (rather than the Inbox) is the point of interest that could address some of the confusions or issues that you raise. For instance, what application is generating the "frobbed" message or at which stack/level is that being raised? Is this a specific issue to LDN senders, receivers, or consumers?

Context matters for LDN content, so, if a sender application is creating the notification (regardless of where the error is coming from), it is deciding for itself as to which human-readable 'inbox' it is referring to. The payload can contain anything (have any model, unless otherwise a constraint is applied), which means that the sender application should include additional information (e.g., something went wrong on the email inbox, something went wrong on the LDN inbox) in order to minimise such potential confusions for the receivers as well as consumers.

It is entirely reasonable to think that implementers may have _poorly_ built applications which don't have the right checks to produce and consume information without confusing the user, but I would suggest that we can have more grounds to consider such changes (i.e., renaming inbox) if we can look at the implementations themselves. Are such cases due to concept 'inbox' or is it something else?

We can also take a look at existing specifications (e.g., https://www.w3.org/TR/notifications/) and tooling out there on the idea of 'notifications'. How did all this work out?

What do you think?

**Comment by csarven** [comment]
Aside: 'container' would be a good alternative to 'inbox', but that'd be describing something a lot broader (container discovery, as opposed to a specific kind of container discovery i.e., inbox). LDP in fact left container discovery out of their spec, and it'd be nice to leave it to them and have them reserve it for possible use. If any of this makes sense in the future, I think the LDP Next CG should tackle that, and perhaps add:

`ldp:inbox rdfs:subPropertyOf ldp:container .`

**Comment by rhiaro** [comment]
I don't think there are any terms we could choose that don't have other meanings already in some related system. (@csarven I think 'container' is  _terrible_ alternative for exactly the reason you describe).

We picked up 'inbox' to match ActivityPub, which is derived from pump.io, which developers have been managing to use for years(?) without confusing users.

@klensin, the error case you describe is a bit high level for me to really understand at which point this error is generated and displayed to the user. If it's generated by the receiver because the sender did something wrong, that would be in the form of an HTTP status code that the sender interprets and feeds back to its UI (ie. "something went wrong with sending this notification, please try again"). If fetching the notification from the receiver by the consumer is what went wrong, then I imagine something along the lines of "could not fetch notification, please try again". But of course it is up to implementations what terminology they use to display errors back to the user. I think implementations explicitly dealing with email inboxes in conjunction with LDN would probably be conscious enough to use the most appropriate terminology ("notification" vs. "inbox") in the right context.  

Perhaps we could add a note in "... Content Considerations" to say implementations should be careful and context-aware about the terminology they present to end users, but it seems also like that should go without saying and/or is none of the spec's business.

**Comment by rhiaro** [comment]
We discussed this in today's WG meeting, and we resolved not to change the name of inbox. The [discussion and resolution in the minutes is here](https://www.w3.org/wiki/Socialwg/2016-10-18-minutes#resolution04). Thanks for your input @klensin, and we hope you feel we've given sufficient consideration to the issue you raised. Apologies if this felt somewhat rushed, we're on a tight time schedule as the WG is due to close at the end of the year!
